kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java

```diff
@@ -44,35 +60,81 @@ class VaultKmsIT {
     private static final String VAULT_TOKEN = "token";
 
     private static final String HASHICORP_VAULT = "hashicorp/vault:1.15";
+    public static final int TLS_PORT = 8202;
+    public static final int PLAIN_PORT = 8200;
     @SuppressWarnings("rawtypes")
-    private VaultContainer vaultContainer;
-    private VaultKms service;
+    private static VaultContainer vaultContainer;
+    private static final String nonTlsService = "nonTls";
+    private static final String pkcs12TrustTlsService = "pkcs12Tls";
+    private static final String filePasswordTlsService = "filePasswordTls";
+    private static final String jksTrustTlsService = "jksTls";
+    private static final String defaultStoreTypeTls = "defaultStoreTypeTls";
+    private static final String insecureService = "tlsInsecure";
+    private static final Map<String, VaultKms> vaultKmsMap = new HashMap<>();
+    private static URI vaultTlsUri;
+
+    private static final CertificateGenerator.Keys keys = CertificateGenerator.generate();
+
+    static List<String> vaultKms() {
+        return vaultKmsMap.keySet().stream().toList();
+    }
 
-    @BeforeEach
-    @SuppressWarnings("resource")
-    void beforeEach() {
+    @BeforeAll
+    public static void beforeAll() {
         assumeThat(DockerClientFactory.instance().isDockerAvailable()).withFailMessage("docker unavailable").isTrue();
+        startVault();
+        String tlsAddress = String.format("https://%s:%s", vaultContainer.getHost(), vaultContainer.getMappedPort(8202));
+        vaultTlsUri = URI.create(tlsAddress);
+        vaultKmsMap.put(nonTlsService, nonTls());
+        vaultKmsMap.put(pkcs12TrustTlsService, getTlsVaultKms(tlsForTrustStoreInlinePassword(keys.pkcs12ClientTruststore())));
+        vaultKmsMap.put(filePasswordTlsService, getTlsVaultKms(tlsForTrustStoreFilePassword(keys.pkcs12ClientTruststore())));
+        vaultKmsMap.put(jksTrustTlsService, getTlsVaultKms(tlsForTrustStoreInlinePassword(keys.jksClientTruststore())));
+        vaultKmsMap.put(defaultStoreTypeTls, getTlsVaultKms(defaultStoreTypeTls(keys.jksClientTruststore())));
+        vaultKmsMap.put(insecureService, getTlsVaultKms(insecureTls()));
+    }
 
+    private static void startVault() {
         vaultContainer = new VaultContainer<>(HASHICORP_VAULT)
                 .withVaultToken(VAULT_TOKEN)
                 .withEnv("VAULT_FORMAT", "json")
+                // in vault server -dev mode the 8200 port cannot be reconfigured to TLS and is also useful so that the
+                // internal vault commands we make while execed into the container can communicate with vault over plain
+                // without having to trust the TLS cert within the container. Plus we can test TLS and non-TLS against the
+                // same container.
+                .withExposedPorts(PLAIN_PORT, TLS_PORT)
```

Comment:
The rational for enabling `-dev` mode makes sense but should we always expose both ports? I guess there is no harm in always exposing the `TLS_PORT` but do we risk false positives in tests if we leave the `PLAIN_PORT` exposed when TLS is enabled?

Reply:
mm, false positives are possible through misconfiguration yes. If you want to exercise TLS and configure the Vault KMS to point at an `http://` endpoint then all the requests will succeed.

Maybe the testing is at the wrong level anyway, I imagine we don't need to test every KMS interaction to check that TLS is configured for each and every one. Maybe this set of tests should run only against one vault configuration and we make a second set of tests that are just for checking TLS through a single API call.

Reply:
pushed up, the VaultKmsIT checks that the KMS business logic and that it's fulfilling some contractual obligations and the VaultKmsTlsIT just wants to check that we can make an API call with various TLS configurations against a vault that only exposes a TLS endpoint (and a negative test showing that we cannot connect without the right configuration).